storage/kv/pebble.go

```diff
@@ -0,0 +1,257 @@
+package kv
+
+import (
+	"context"
+
+	"github.com/cockroachdb/pebble"
+	"github.com/pkg/errors"
+	"github.com/ssvlabs/ssv/storage/basedb"
+	"go.uber.org/zap"
+)
+
+// TODO: reconsider prefixes?
+var _ basedb.Database = &PebbleDB{}
+
+// PebbleDB struct
+type PebbleDB struct {
+	logger *zap.Logger
+
+	db *pebble.DB
+
+	ctx    context.Context
+	cancel context.CancelFunc
+	//wg     sync.WaitGroup
+	//
+	//// gcMutex is used to ensure that only one GC cycle is running at a time.
+	//gcMutex sync.Mutex
+}
+
+func NewPebbleDB(pCtx context.Context, logger *zap.Logger, path string, opts *pebble.Options) (*PebbleDB, error) {
+	ctx, cancel := context.WithCancel(pCtx)
+	pbdb := &PebbleDB{
+		logger: logger,
+		ctx:    ctx,
+		cancel: cancel,
+	}
+	pdb, err := pebble.Open(path, opts)
+	if err != nil {
+		return nil, err
+	}
+	pbdb.db = pdb
+	return pbdb, nil
+}
+
+func (pdb *PebbleDB) Close() error {
+	pdb.cancel()
+	return pdb.db.Close()
+}
+
+func (pdb *PebbleDB) Get(prefix []byte, key []byte) (basedb.Obj, bool, error) {
+
+	b, closer, err := pdb.db.Get(append(prefix, key...))
+	if errors.Is(err, pebble.ErrNotFound) {
+		return basedb.Obj{}, false, nil
+	}
+	if err != nil {
+		return basedb.Obj{}, false, err
+	}
+	// PebbleDB returned slice is valid until closer.Close() is called
+	// hence we copy
+	//
+	defer func() { _ = closer.Close() }()
+
+	out := make([]byte, len(b))
+	copy(out, b)
+	return basedb.Obj{Key: key, Value: out}, true, nil
+}
+
+func (pdb *PebbleDB) Set(prefix, key, value []byte) error {
+	return pdb.db.Set(append(prefix, key...), value, pebble.Sync)
+}
+
+func (pdb *PebbleDB) Delete(prefix, key []byte) error {
+
+	return pdb.db.Delete(key, pebble.Sync)
+}
+
+func (pdb *PebbleDB) GetMany(prefix []byte, keys [][]byte, iterator func(basedb.Obj) error) error {
+	for _, key := range keys {
+		// Create a new slice for the full key (prefix + key)
+		fullKey := make([]byte, len(prefix)+len(key))
+		copy(fullKey, prefix)
+		copy(fullKey[len(prefix):], key)
+
+		// Retrieve the value from Pebble.
+		value, closer, err := pdb.db.Get(fullKey)
+		if err != nil {
+			// If the key isn't found, skip it.
+			if errors.Is(err, pebble.ErrNotFound) {
+				continue
+			}
+			return err
+		}
+
+		// Since the returned value is only valid until closer.Close(),
+		// we make a copy of it.
+		vcopy := make([]byte, len(value))
+		copy(vcopy, value)
+
+		// Close the closer to release resources.
+		if err := closer.Close(); err != nil {
+			return err
+		}
+
+		// Wrap the key/value in an object satisfying basedb.Obj.
+		obj := basedb.Obj{
+			Key:   fullKey,
+			Value: vcopy,
+		}
+
+		// Call the iterator callback.
+		if err := iterator(obj); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func makePrefixIter(dbOrBatch pebble.Reader, prefix []byte) (*pebble.Iterator, error) {
```

Comment:
Can you please add some comments on why we need the incremental logic in this function? Also, I think `keyUpperBound` needs unit tests